File: android/app/src/main/java/market/foodhome/app/recovery/CrashLoopBreaker.kt
package market.foodhome.app.recovery

import java.time.Duration
import java.time.Instant
import java.util.ArrayDeque

class CrashLoopBreaker(
    private val window: Duration = Duration.ofMinutes(1),
) {
    private val crashes = ArrayDeque<Instant>()
    private var activeRouteKey: String? = null

    @Synchronized
    fun recordCrash(routeKey: String, now: Instant): Boolean {
        if (activeRouteKey != routeKey) {
            crashes.clear()
            activeRouteKey = routeKey
        }
        crashes.addLast(now)
        val cutoff = now.minus(window)
        while (crashes.firstOrNull()?.isBefore(cutoff) == true) crashes.removeFirst()
        return crashes.size >= 2
    }

    @Synchronized
    fun reset() {
        crashes.clear()
        activeRouteKey = null
    }
}

File: ios/FoodHomeApp/Recovery/CrashLoopBreaker.swift
import Foundation

final class CrashLoopBreaker {
    private let window: TimeInterval
    private var crashes: [Date] = []
    private var activeRouteKey: String?

    init(window: TimeInterval = 60) {
        self.window = window
    }

    func recordCrash(routeKey: String, at now: Date) -> Bool {
        if activeRouteKey != routeKey {
            crashes.removeAll()
            activeRouteKey = routeKey
        }
        crashes.append(now)
        let cutoff = now.addingTimeInterval(-window)
        crashes.removeAll { $0 < cutoff }
        return crashes.count >= 2
    }

    func reset() {
        crashes.removeAll()
        activeRouteKey = nil
    }
}
